### PR TITLE
Make `BlockImport` and `Verifier` async

### PR DESCRIPTION
**Pull request by bkchr**
This pr makes both traits `BlockImport` and `Verifier` async traits. This change enables us to use futures in block import and verifier.

polkadot companion: https://github.com/paritytech/polkadot/pull/2759

### CODE REVIEW

**Review by lovelaced — APPROVED** [review]
approving on behalf of CI

File: client/consensus/common/src/shared_data.rs [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+		Self {
+			inner: Arc::new(Mutex::new(SharedDataInner { shared_data, locked: false })),
+			cond_var: Default::default(),
+		}
+	}
+
+	/// Acquire access to the shared data.
+	///
+	/// This will give mutable access to the shared data. After the returned mutex guard is dropped,
+	/// the shared data is accessable by other threads. So, this function should be used when
```
Comment (andresilva) [inline client/consensus/common/src/shared_data.rs:196]:
```suggestion
	/// the shared data is accessible by other threads. So, this function should be used when
```

File: client/consensus/babe/src/tests.rs [hunk 2]
```diff
@@ -405,8 +418,8 @@ fn run_one_test(
 			// run each future until we get one of our own blocks with number higher than 5
 			// that was produced locally.
 			client.import_notification_stream()
-				.take_while(move |n| future::ready(n.header.number() < &5 || {
-					if n.origin == BlockOrigin::Own {
+				.take_while(move |n| future::ready(dbg!(n.header.number()) < &5 || {
```
Comment (andresilva) [inline client/consensus/babe/src/tests.rs:421]:
```suggestion
				.take_while(move |n| future::ready(n.header.number() < &5 || {
```

File: client/consensus/babe/src/tests.rs [hunk 3]
```diff
@@ -405,8 +418,8 @@ fn run_one_test(
...
+					if dbg!(n.origin) == BlockOrigin::Own {
```
Comment (andresilva) [inline client/consensus/babe/src/tests.rs:422]:
```suggestion
					if n.origin == BlockOrigin::Own {
```

File: client/finality-grandpa/src/import.rs [hunk 4]
```diff
@@ -315,7 +320,7 @@ where
 		let is_descendent_of = is_descendent_of(&*self.inner, Some((hash, parent_hash)));
 
 		let mut guard = InnerGuard {
-			guard: Some(self.authority_set.inner().write()),
+			guard: Some(self.authority_set.inner_locked()),
```
Comment (andresilva) [inline client/finality-grandpa/src/import.rs:323]: Technically we don't need to lock here as we won't be releasing the outer lock (there's no await boundary below).

File: client/consensus/common/src/shared_data.rs [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+	inner: MutexGuard<'a, SharedDataInner<T>>,
+	/// The [`SharedData`] instance that created this instance.
+	///
+	/// This instance is only taken on drop or when calling [`Self::release_mutex`].
+	shared_data: Option<SharedData<T>>,
+}
+
+impl<'a, T> SharedDataLocked<'a, T> {
+	/// Release the mutex, but keep the shared data locked.
+	pub fn release_mutex(mut self) -> SharedDataLockedUpgradable<T> {
```
Comment (andresilva) [inline client/consensus/common/src/shared_data.rs:76]: It would be nice to get a better name for this, but I have no suggestions 🙈

**Review by andresilva — APPROVED** [review]
lgtm, just minor nits.

File: client/finality-grandpa/src/import.rs (hunk 4 above)
Comment (bkchr) [reply client/finality-grandpa/src/import.rs:323]: Yeah, not here, but in the outer context that is calling this function and that requires that the set stays locked.

**Comment by bkchr** [comment]
bot merge

**Pull request merged**